core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java

```diff
@@ -78,7 +84,24 @@ public SqlItemOperator(String name,
       SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
     call.operand(0).unparse(writer, leftPrec, 0);
     final SqlWriter.Frame frame = writer.startList("[", "]");
-    call.operand(1).unparse(writer, 0, 0);
+    if (offset == 0) {
+      // Technically these functions are extensions to standard SQL,
+      // however, the standard SQL parser accepts them,
+      // so there must be a way to unparse them.
+      String function = safe ? "SAFE_OFFSET" : "OFFSET";
```

Comment:
If we have `String funcId`, we can write the code 
```
if (funcId != null) {
      final SqlWriter.Frame offsetFrame = writer.startFunCall(funcId);
      call.operand(1).unparse(writer, 0, 0);
      writer.endFunCall(offsetFrame);
}
else {
      call.operand(1).unparse(writer, 0, 0);
}
```

Reply:
This is much better, thank you.